src/core/btree_set.h

```diff
@@ -0,0 +1,281 @@
+    std::less<T> cmp;
+    return cmp(a, b) ? -1 : (cmp(b, a) ? 1 : 0);
+  }
+};
+
+template <typename T> struct BTreePolicy {
+  using Key_t = T;
+  using KeyCompareTo = DefaultCompareTo<T>;
+};
+
+template <typename T, typename Policy = BTreePolicy<T> > class BTree {
+  BTree(const BTree&) = delete;
+  BTree& operator=(const BTree&) = delete;
+
+  using BTreeNode = detail::BTreeNode<Policy>;
+  using BTreePath = detail::BTreePath<Policy>;
+
+ public:
+  using Key_t = typename Policy::Key_t;
+
+  BTree(PMR_NS::memory_resource* mr = PMR_NS::get_default_resource()) : mr_(mr) {
+  }
+
+  ~BTree() {
+    Clear();
+  }
+
+  // true if inserted, false if skipped.
+  bool Insert(Key_t item);
+
+  bool Contains(Key_t item) const;
+
+  size_t Height() const {
+    return height_;
+  }
+
+  size_t Size() const {
+    return count_;
+  }
+
+  size_t NodeCount() const {
+    return num_nodes_;
+  }
+
+  void Clear();
+
+  BTreeNode* DEBUG_root() {
+    return root_;
+  }
+
+ private:
+  BTreeNode* CreateNode(bool leaf);
+
+  void DestroyNode(BTreeNode* node);
+
+  // Unloads the full leaf to allow insertion of additional item.
+  // The leaf should be the last one in the path.
+  std::pair<BTreeNode*, Key_t> InsertFullLeaf(Key_t item, const BTreePath& path);
+
+  // Charts the path towards key. Returns true if key is found.
+  // In that case path->Last().first->Key(path->Last().second) == key.
+  // Fills the tree path not including the key itself.
+  bool Locate(Key_t key, BTreePath* path) const;
+
+  BTreeNode* root_ = nullptr;  // root node or NULL if empty tree
+  uint32_t count_ = 0;         // number of items in tree
+  uint32_t height_ = 0;        // height of tree from root to leaf
+  uint32_t num_nodes_ = 0;     // number of nodes in tree
+  PMR_NS::memory_resource* mr_;
+};
+
+template <typename T, typename Policy> bool BTree<T, Policy>::Contains(Key_t item) const {
+  BTreePath path;
+  bool found = Locate(item, &path);
+  return found;
+}
+
+template <typename T, typename Policy> void BTree<T, Policy>::Clear() {
+  if (!root_)
+    return;
+
+  BTreePath path;
+  BTreeNode* node = root_;
+
+  auto deep_left = [&](unsigned pos) {
+    do {
+      path.Push(node, pos);
+      node = node->Child(pos);
+      pos = 0;
+    } while (!node->IsLeaf());
+  };
+
+  if (!root_->IsLeaf())
+    deep_left(0);
+
+  while (true) {
+    DestroyNode(node);
+
+    if (path.Depth() == 0) {
+      break;
+    }
+    node = path.Last().first;
+    unsigned pos = path.Last().second;
+    path.Pop();
+    if (pos < node->NumItems()) {
+      deep_left(pos + 1);
+    }
+  }
+  root_ = nullptr;
+  height_ = count_ = 0;
+}
+
+template <typename T, typename Policy> bool BTree<T, Policy>::Insert(Key_t item) {
+  if (!root_) {
+    root_ = CreateNode(true);
+    root_->InitSingle(item);
+    count_ = height_ = 1;
+
+    return true;
+  }
+
+  BTreePath path;
+  bool found = Locate(item, &path);
+
+  if (found) {
+    return false;
+  }
+
+  assert(path.Depth() > 0u);
```

Comment:
raw assert to avoid being dependant on glog?

Reply:
yes, I do not like having glog in header files

Reply:
this is why I like explicit instantiations 😆

Reply:
I do not mind it. If @chakaz  does not have strong objections - I can move this to cc file.

Reply:
I always prefer implementations to be in the .cc, like.. *always* :)

Reply:
Obviously except for when it's impossible, like with templates :(